src/ray/util/container_util.h

```diff
@@ -28,56 +30,92 @@
 namespace ray {
 
 template <typename T>
-std::string debug_string(const T &obj) {
-  std::stringstream ss;
-  ss << obj;
-  return ss.str();
+class DebugStringWrapper;
+
+// The actual interface.
+template <typename T>
+DebugStringWrapper<T> debug_string(const T &t) {
+  return DebugStringWrapper<T>(t);
+}
+
+/// Wrapper for `debug_string(const T&)`.
+template <typename T>
+class DebugStringWrapper {
+ public:
+  explicit DebugStringWrapper(const T &obj) : obj_(obj) {}
+
+  // Only initialized for the blessed container types listed below with operator<<
+  // specializations.
+  std::ostream &_container_debug_string_impl(std::ostream &os) const {
+    os << "[";
+    for (auto it = obj_.begin(); it != obj_.end(); ++it) {
+      if (it != obj_.begin()) {
+        os << ", ";
+      }
+      os << debug_string(*it);
+    }
+    os << "]";
+    return os;
+  }
+
+  // Public but OK, since it's const &.
```

Comment:
But I would still prefer making it provide; for example, people could use `const_cast` for mutability.

Reply:
I don't think it's a big issue, or maybe we use `struct` and discard `public`?

Reply:
should not be a big issue - even if you have a getter method that returns cosnt&, caller can still do const_cast. It'd be the caller's fault to mess up things.